src/main/java/org/sonatype/nexus/repository/composer/internal/ComposerHostedMetadataFacetImpl.java

```diff
@@ -0,0 +1,121 @@
+ * Eclipse Foundation. All other trademarks are the property of their respective owners.
+ */
+package org.sonatype.nexus.repository.composer.internal;
+
+import java.io.IOException;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+
+import org.sonatype.nexus.common.event.EventManager;
+import org.sonatype.nexus.common.stateguard.Guarded;
+import org.sonatype.nexus.repository.FacetSupport;
+import org.sonatype.nexus.repository.storage.Asset;
+import org.sonatype.nexus.repository.storage.AssetCreatedEvent;
+import org.sonatype.nexus.repository.storage.AssetDeletedEvent;
+import org.sonatype.nexus.repository.storage.AssetEvent;
+import org.sonatype.nexus.repository.storage.AssetUpdatedEvent;
+import org.sonatype.nexus.repository.storage.StorageFacet;
+import org.sonatype.nexus.transaction.UnitOfWork;
+
+import com.google.common.eventbus.AllowConcurrentEvents;
+import com.google.common.eventbus.Subscribe;
+import org.joda.time.DateTime;
+
+import static jline.internal.Preconditions.checkNotNull;
+import static org.joda.time.DateTime.now;
+import static org.sonatype.nexus.common.stateguard.StateGuardLifecycleSupport.State.STARTED;
+import static org.sonatype.nexus.repository.composer.internal.ComposerAttributes.P_PROJECT;
+import static org.sonatype.nexus.repository.composer.internal.ComposerAttributes.P_VENDOR;
+import static org.sonatype.nexus.repository.storage.AssetEntityAdapter.P_ASSET_KIND;
+
+/**
+ * Concrete implementation of {@code ComposerHostedMetadataFacet} using events. Essentially a stripped-down form of the
+ * mechanism behind the Yum createrepo implementation in our internal codebase.
+ */
+@Named
+public class ComposerHostedMetadataFacetImpl
+    extends FacetSupport
+    implements ComposerHostedMetadataFacet
+{
+  private final EventManager eventManager;
+
+  @Inject
+  public ComposerHostedMetadataFacetImpl(final EventManager eventManager)
+  {
+    this.eventManager = checkNotNull(eventManager);
+  }
+
+  @Subscribe
+  @Guarded(by = STARTED)
+  @AllowConcurrentEvents
+  public void on(final AssetDeletedEvent deleted) {
+    if (matchesRepository(deleted) && isEventRelevant(deleted)) {
+      invalidateMetadata(deleted);
+    }
+  }
+
+  @Subscribe
+  @Guarded(by = STARTED)
+  @AllowConcurrentEvents
+  public void on(final AssetCreatedEvent created) {
+    if (matchesRepository(created) && isEventRelevant(created)) {
+      invalidateMetadata(created);
+    }
+  }
+
+  @Subscribe
+  @Guarded(by = STARTED)
+  @AllowConcurrentEvents
+  public void on(final AssetUpdatedEvent updated) {
+    if (matchesRepository(updated) && isEventRelevant(updated) && hasBlobBeenUpdated(updated)) {
+      invalidateMetadata(updated);
+    }
+  }
+
+  @Subscribe
+  @Guarded(by = STARTED)
+  public void on(final ComposerHostedMetadataInvalidationEvent event) throws IOException {
+    if (getRepository().getName().equals(event.getRepositoryName())) {
```

Comment:
can you use `matchesepository(event)` here instead

Reply:
> can you use matchesepository(event) here instead

Nope, there's no shared superclass, and I doubt we want to introduce a new interface to handle this (nor could we without touching the main NXRM codebase). So in this case it's more of a typing issue than anything else (I'm not sure how we could imagine this being an `AssetEvent`, for example).